File: core/ucp/apps/ucp-api/src/ports/outbound/project.provider.ts
import { ZitadelRole, ZitadelUser } from '../../domain/dtos/zitadel.dto.js';

export const PROJECT_PROVIDER = Symbol('PROJECT_PROVIDER');

export interface IProjectProvider {
  createProjectGrant(
    orgId: string,
    projectId: string,
    roleKeys: string[],
  ): Promise<unknown>;
  deleteProjectGrant(orgId: string, projectId: string): Promise<void>;
  getRoles(): Promise<ZitadelRole[]>;
  getUsers(orgId: string): Promise<ZitadelUser[]>;
}

File: core/ucp/apps/ucp-api/src/ucp_api/ports/outbound/project_provider.py
from typing import Protocol, List
from ucp_api.domain.dtos.zitadel_dtos import ZitadelRole, ZitadelUser

class IProjectProvider(Protocol):
    async def create_project_grant(self, org_id: str, project_id: str, role_keys: List[str]) -> None:
        """Grants a project to an organization with specific roles"""
        ...
        
    async def delete_project_grant(self, org_id: str, project_id: str) -> None:
        """Removes a project grant from an organization"""
        ...
        
    async def get_roles(self) -> List[ZitadelRole]:
        """Gets all roles for the UCP project"""
        ...
        
    async def get_users(self, org_id: str) -> List[ZitadelUser]:
        """Gets all users with access to this project in a specific org"""
        ...
